### PR TITLE
Allow separate configuration of ELB subnets

### PR DESCRIPTION
**Pull request by izaaklauer**
Prior to this PR, you could specify a set of subnets that would be used for the ALB and the ECS service. A common vpc configuration though is to put ALBs in public subnets, and keep app instances isolated in private subnets behind a nat gateway.

To facilitate that configuration, this PR adds a new `subnets` config field under the `ALB` config. Internally, it also splits the `subnet` resources into two separate `serviceSubnets` and `albSubnets` resources.

Fixes https://github.com/hashicorp/waypoint/issues/2161

### AssignPublicIP

One loosely-related change included here too - one main reason to have app instances in a private subnet is to make it harder for the internet to ingress directly to the app. Prior to this change, all ECS instances were given public IPs, which violates the spirit of isolating the apps in a private subnet.

To fix that, this change also introduces an "assign_public_ip" field, which was commented-out before. It defaults to true, because instances in a subnet without a nat gateway (including the default subnets when not configured otherwise) need it to egress to public ECR to pull their containers.

In the future, we could notice when `assign_public_ip` is unset, check the configuration of the specified subnets to see if its route table includes a default route to a nat gateway, and only assign a public IP if it does not.

### CODE REVIEW

File: builtin/aws/ecs/platform.go [hunk 1]
```diff
@@ -2389,14 +2442,16 @@ type Config struct {
...
-	// set this to false in the default subnets, ECS can't pull the image. Leaving
-	// it disabled until we figure out how to handle that onramp case.
-	// AssignPublicIp bool `hcl:"assign_public_ip,optional"`
+	// Assign each task a public IP. Default true.
+	// Note: If private subnets have been specified for ECS tasks, and
+	// no NAT gateway is configured, ECS will be unable to pull your image
+	// from ECR and your services will be unable to start.
+	// NOTE(izaak): If this is not set, we can probably set it to true if
+	// the ecs subnet's route table has a nat gw default route, and true if they don't.
+	AssignPublicIp *bool `hcl:"assign_public_ip,optional"`
```
Comment (izaaklauer) [inline builtin/aws/ecs/platform.go:2454]: Using a *bool so we can potentially set a clever default in the future if this value is unset.

**Review by evanphx — CHANGES_REQUESTED** [review]
Just a couple of minor things.

File: builtin/aws/ecs/platform.go [hunk 2]
```diff
@@ -2732,6 +2797,12 @@ deploy {
 		docs.Default("3000"),
 	)
 
+	doc.SetField(
+		"assign_public_ip",
+		"assign a public ip address to tasks. Defaults to true. Ignored if using an ec2 cluster.",
+		docs.Default("true"),
```
[the remark this answers is not in the document]
Comment (izaaklauer) [reply builtin/aws/ecs/platform.go:2803]: Done!

File: builtin/aws/ecs/platform.go [hunk 3]
```diff
@@ -207,13 +207,24 @@ func (p *Platform) resourceManager(log hclog.Logger, dcr *component.DeclaredReso
...
 		)),
 		resource.WithResource(resource.NewResource(
-			resource.WithName("subnets"),
+			resource.WithName("service subnets"),
+			resource.WithType("subnets"),
 			resource.WithPlatform(platformName),
-			resource.WithState(&Resource_Subnets{}),
-			resource.WithCreate(p.resourceSubnetsDiscover),
+			resource.WithState(&Resource_ServiceSubnets{}),
+			resource.WithCreate(p.resourceServiceSubnetsDiscover),
```
[the remark this answers is not in the document]
Comment (izaaklauer) [reply builtin/aws/ecs/platform.go:214]: The subnet resources haven't showed up in any released version yet - I only merged it into main the other day. So it's probably backwards incompatible with main, but not for any of our users.

**Review by briancain — APPROVED** [review]
Nice! 👍🏻 Makes sense

**Pull request merged**